File: rplugin/python3/curstr/action/cursor.py
import re
from typing import Tuple

from neovim import Nvim

from curstr.echoable import Echoable
from curstr.info import SourceExecuteInfo


class Cursor(Echoable):
    def __init__(self, vim: Nvim, info: SourceExecuteInfo) -> None:
        self._vim = vim
        self._info = info

    def target_option(func):
        def wrapper(self, *args, **kwargs) -> str:
            target_string = self._info.string
            if target_string:
                return target_string
            # FIXME: ignore mypy error: "Cursor" not callable
            # mypy #1927 ?
            return func(self, *args, **kwargs)  # type: ignore

        return wrapper

    @target_option  # type: ignore
    def get_word(self, added_iskeyword="") -> str:
        if added_iskeyword == "":
            return self._vim.call("expand", "<cword>")

        added = ",".join(added_iskeyword)
        try:
            self._vim.command("setlocal iskeyword+={}".format(added))
            word = self._vim.call("expand", "<cword>")
        finally:
            self._vim.command("setlocal iskeyword-={}".format(added))

        return word

    @target_option  # type: ignore
    def get_file_path(self, added_isfname="") -> str:
        if added_isfname == "":
            return self._vim.call("expand", "<cfile>")

        added = ",".join(added_isfname)
        try:
            self._vim.command("setlocal isfname+={}".format(added))
            file_path = self._vim.call("expand", "<cfile>")
        finally:
            self._vim.command("setlocal isfname-={}".format(added))

        return file_path

    def get_file_path_with_position(
        self, added_isfname=""
    ) -> Tuple[str, Tuple[int, int]]:
        file_path = self.get_file_path(added_isfname)

        cword = self._vim.call("expand", "<cWORD>")
        pattern = "{}:(\\d+)(,\\d+)?".format(file_path)
        match = re.match(pattern, cword)
        if match is None:
            return (file_path, (-1, -1))

        row = int(match.group(1))
        if match.group(2) is None:
            return (file_path, (row, 1))

        column = int(match.group(2)[1:])
        position = (row, column)
        return (file_path, position)

    def get_word_with_range(self, char_pattern="\\k") -> Tuple[str, Tuple[int, int]]:
        current_pos = self._vim.current.window.cursor
        line = self._vim.current.line
        word, start_byte = self._vim.call(
            "matchstrpos",
            line,
            "\\v[{}]*%{}c[{}]+".format(char_pattern, current_pos[1] + 1, char_pattern),
        )[:2]

        if start_byte == -1:
            return ("", (-1, -1))

        after_part = self._vim.call("strpart", line, start_byte)
        start = len(line) - len(after_part)
        end = start + len(word)
        word_range = (start, end)

        return (word, word_range)

    @target_option  # type: ignore
    def get_line(self) -> str:
        line = self._vim.current.line
        return line

File: lua/curstr/core/cursor.lua
local M = {}

function M.word(_, added_iskeyword)
  if added_iskeyword == nil then
    return vim.fn.expand("<cword>")
  end

  local origin_iskeyword = vim.bo.iskeyword
  local splitted = vim.split(origin_iskeyword, ",", { plain = true })
  vim.list_extend(splitted, vim.split(added_iskeyword, "", { plain = true }))
  local new_iskeyword = table.concat(splitted, ",")

  vim.bo.iskeyword = new_iskeyword
  local word = vim.fn.expand("<cword>")
  vim.bo.iskeyword = origin_iskeyword

  return word
end

function M.word_with_range(_, char_pattern)
  char_pattern = char_pattern or "\\k"
  local pos = vim.api.nvim_win_get_cursor(0)
  local line = vim.api.nvim_get_current_line()
  local pattern = ("\\v[%s]*%%%sc[%s]+"):format(char_pattern, pos[2] + 1, char_pattern)
  local word, start_byte = unpack(vim.fn.matchstrpos(line, pattern))
  if start_byte == -1 then
    return "", nil
  end
  local after_part = vim.fn.strpart(line, start_byte)
  local s = #line - #after_part
  local e = s + #word
  local word_range = { s, e }
  return word, word_range
end

local trim_prefix = function(file_path)
  file_path = file_path:gsub("^file://", "")
  return file_path
end

function M.file_path(_, added_isfname)
  if added_isfname == nil then
    return trim_prefix(vim.fn.expand("<cfile>"))
  end

  local origin_isfname = vim.o.isfname
  local splitted = vim.split(origin_isfname, ",", { plain = true })
  vim.list_extend(splitted, vim.split(added_isfname, "", { plain = true }))
  local new_isfname = table.concat(splitted, ",")

  vim.o.isfname = new_isfname
  local path = vim.fn.expand("<cfile>")
  vim.o.isfname = origin_isfname

  return trim_prefix(path)
end

local parse_position = function(str, s, e)
  local position = str:sub(s + 1, e)
  local row, col = unpack(vim.split(position, "[,:]"))
  return { tonumber(row), tonumber(col or 1) }
end

local file_path_regex = vim.regex("\\v[^:]+:\\zs(\\d+)([,:]\\d+)?")

function M.file_path_with_position(_, added_isfname)
  local file_path = M.file_path(added_isfname)

  do
    local s, e = file_path_regex:match_str(file_path)
    if s ~= nil then
      local path = file_path:sub(1, s - 1)
      local position = parse_position(file_path, s, e)
      return path, position
    end
  end

  local cword = trim_prefix(vim.fn.expand("<cWORD>"))
  local pattern = ("\\v%s:\\zs(\\d+)([,:]\\d+)?"):format(file_path)
  local ok, regex = pcall(vim.regex, pattern)
  if not ok then
    return file_path, nil
  end
  local s, e = regex:match_str(cword)
  if s == nil then
    return file_path, nil
  end
  return file_path, parse_position(cword, s, e)
end

function M.line_with_range()
  local line = vim.api.nvim_get_current_line()
  return line, { 1, #line }
end

return M
